base/common/src/main/java/com/netscape/certsrv/cert/CertEnrollmentRequest.java

```diff
@@ -349,18 +360,225 @@ public boolean equals(Object obj) {
+
+        Element attributesElement = document.createElement("Attributes");
+
+        for (Map.Entry<String, String> attribute : getAttributes().entrySet()) {
+            Element attributeElement = document.createElement("Attribute");
+            attributeElement.appendChild(document.createTextNode(attribute.getValue()));
+            attributeElement.setAttribute("name", attribute.getKey());
+            attributesElement.appendChild(attributeElement);
+
+        }
+        certEnrollmentRequestElement.appendChild(attributesElement);
+
+        if (getProfileId() != null && !getProfileId().isEmpty()) {
+            Element profileIdElement = document.createElement("ProfileID");
+            profileIdElement.appendChild(document.createTextNode(getProfileId()));
+            certEnrollmentRequestElement.appendChild(profileIdElement);
+        }
+
+        if (getServerSideKeygenP12Passwd() != null && !getServerSideKeygenP12Passwd().isEmpty()) {
+            Element getServerSideKeygenP12PasswdElement = document.createElement("ServerSideKeygenP12Passwd");
+            getServerSideKeygenP12PasswdElement.appendChild(document.createTextNode(getServerSideKeygenP12Passwd()));
+            certEnrollmentRequestElement.appendChild(getServerSideKeygenP12PasswdElement);
+        }
+
+        Element renewalElement = document.createElement("Renewal");
+        renewalElement.appendChild(document.createTextNode(String.valueOf(isRenewal())));
+        certEnrollmentRequestElement.appendChild(renewalElement);
+
+        if (getSerialNum() != null) {
+            Element serialNumberElement = document.createElement("SerialNumber");
+            serialNumberElement.appendChild(document.createTextNode(getSerialNum().toHexString()));
+            certEnrollmentRequestElement.appendChild(serialNumberElement);
+        }
+
+        if (getRemoteHost() != null) {
+            Element remoteHostElement = document.createElement("RemoteHost");
+            remoteHostElement.appendChild(document.createTextNode(getRemoteHost()));
+            certEnrollmentRequestElement.appendChild(remoteHostElement);
+        }
+
+        if (getRemoteAddr() != null) {
+            Element remoteAddressElement = document.createElement("RemoteAddress");
+            remoteAddressElement.appendChild(document.createTextNode(getRemoteAddr()));
+            certEnrollmentRequestElement.appendChild(remoteAddressElement);
+        }
+
+        for (ProfileInput input: getInputs()) {
+            certEnrollmentRequestElement.appendChild(input.toDOM(document));
+        }
+
+        for (ProfileOutput output: getOutputs()) {
+
+            certEnrollmentRequestElement.appendChild(output.toDOM(document));
+
+            NodeList nodes = certEnrollmentRequestElement.getElementsByTagName("profileOutput");
+            System.out.println(nodes.getLength());
+            for (int i=0; i < nodes.getLength(); i++) {
+                document.renameNode(nodes.item(i), null, "Output");
+            }
```

Comment:
That should work, so feel free to merge (but remove the `println()` first).

If you want, you probably could refactor `ProfileOutput.toDOM()` like this:
```
public void toDOM(Document document, Element element) {
    element.setAttribute("id", id);
    ...
}

public Element toDOM(Document document) {
    Element poElement = document.createElement("profileOutput");
    toDOM(document, poElement);
    return poElement;
}
```
Then you can call it like this:
```
Element output = document.createElement("Output");
output.toDOM(document, output);
createEnrollmentRequestElement.appendChild(output);
```

Reply:
Updated. Thanks that looks a lot cleaner!